{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "from sqlalchemy.orm import sessionmaker\n",
    "from sqlalchemy import create_engine\n",
    "import mysql.connector\n",
    "import pysam\n",
    "import pandas as pd\n",
    "# from model import db, Patient, Variant, Gene, patient_has_variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Conexi칩n exitosa.\n"
     ]
    }
   ],
   "source": [
    "db_connection = mysql.connector.connect(\n",
    "    host=\"localhost\",       # Cambia si est치s usando un servidor remoto\n",
    "    user=\"root\",            # Tu usuario de MySQL\n",
    "    password=\"1234\",  # Cambia a tu contrase침a\n",
    "    database=\"variantes_prueba\"  # Nombre de la base de datos donde tienes las variantes\n",
    ")\n",
    "cursor = db_connection.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GBM and LUAD don't show any mutation matching. \n",
    "vcf = pysam.VariantFile(\"/home/marti/MBHS/DBW/OncoTrack/vcf_mutations/LUAD_sample_muts.hg38.vcf\")\n",
    "variants = []\n",
    "for record in vcf:\n",
    "    chrom = record.chrom \n",
    "    pos = record.pos  \n",
    "    ref = record.ref  \n",
    "    alt = ','.join(str(a) for a in record.alts)\n",
    "    \n",
    "    for alt in record.alts:\n",
    "        if len(ref) > 1 or any(len(a) > 1 for a in record.alts):\n",
    "            ref = ref[1:] if len(ref) > 1 else \"\"\n",
    "            alt = alt[1:] if len(alt) > 1 else \"\"\n",
    "\n",
    "        variants.append({\n",
    "            \"chrom\": chrom,\n",
    "            \"pos\": pos,\n",
    "            \"ref\": ref,\n",
    "            \"alt\": alt\n",
    "        })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_8842/1191615146.py:7: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('/home/marti/MBHS/DBW/OncoTrack/cosmut_db.tsv', sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1156314"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sqlalchemy import create_engine\n",
    "import pandas as pd\n",
    "\n",
    "# Conexi칩n con MySQL\n",
    "# engine = create_engine('mysql+mysqlconnector://root:1234@localhost/variantes_prueba')\n",
    "\n",
    "# df = pd.read_csv('/home/marti/MBHS/DBW/OncoTrack/cosmut_db.tsv', sep='\\t')\n",
    "\n",
    "# df.to_sql('cosmic_mutations', con=engine, if_exists='replace', index=False)\n",
    "\n",
    "# CREATE INDEX idx_chromosome ON cosmic_mutations(CHROMOSOME(100));\n",
    "# CREATE INDEX idx_genome_start ON cosmic_mutations(GENOME_START);\n",
    "# CREATE INDEX idx_genomic_wt_allele ON cosmic_mutations(GENOMIC_WT_ALLELE(100)); \n",
    "# CREATE INDEX idx_genomic_mut_allele ON cosmic_mutations(GENOMIC_MUT_ALLELE(100));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'gene_sym': 'KRAS', 'chrom': '12', 'pos': 25245350, 'ref': 'C', 'alt': 'T', 'cosmic_id': 'COSV55497369', 'mutation_desc': 'missense_variant'}\n",
      "{'gene_sym': 'ACVR1B', 'chrom': '12', 'pos': 51994055, 'ref': 'T', 'alt': 'C', 'cosmic_id': 'COSV99231668', 'mutation_desc': 'missense_variant'}\n",
      "{'gene_sym': 'ACVR1B', 'chrom': '12', 'pos': 51994061, 'ref': 'T', 'alt': 'G', 'cosmic_id': 'COSV99231669', 'mutation_desc': 'missense_variant'}\n",
      "{'gene_sym': 'BRCA2', 'chrom': '13', 'pos': 32339600, 'ref': 'T', 'alt': 'C', 'cosmic_id': 'COSV101204513', 'mutation_desc': 'missense_variant'}\n",
      "{'gene_sym': 'CDH11', 'chrom': '16', 'pos': 64988180, 'ref': 'C', 'alt': 'A', 'cosmic_id': 'COSV99219293', 'mutation_desc': 'stop_gained'}\n",
      "{'gene_sym': 'KLK2', 'chrom': '19', 'pos': 50878638, 'ref': 'G', 'alt': 'T', 'cosmic_id': 'COSV100320022', 'mutation_desc': '3_prime_UTR_variant'}\n",
      "{'gene_sym': 'NCOA1', 'chrom': '2', 'pos': 24707381, 'ref': 'G', 'alt': 'T', 'cosmic_id': 'COSV99946683', 'mutation_desc': 'missense_variant'}\n",
      "{'gene_sym': 'LRP1B', 'chrom': '2', 'pos': 141013603, 'ref': 'C', 'alt': 'T', 'cosmic_id': 'COSV101258667', 'mutation_desc': 'missense_variant'}\n",
      "{'gene_sym': 'CNTNAP2', 'chrom': '7', 'pos': 147485934, 'ref': 'G', 'alt': 'T', 'cosmic_id': 'COSV100668890', 'mutation_desc': 'splice_acceptor_variant'}\n",
      "{'gene_sym': 'NBN', 'chrom': '8', 'pos': 89981414, 'ref': 'C', 'alt': 'A', 'cosmic_id': 'COSV99619849', 'mutation_desc': 'missense_variant'}\n",
      "{'gene_sym': 'FAM135B', 'chrom': '8', 'pos': 138151523, 'ref': 'A', 'alt': 'T', 'cosmic_id': 'COSV99425921', 'mutation_desc': 'synonymous_variant'}\n",
      "{'gene_sym': 'JAK2', 'chrom': '9', 'pos': 5080377, 'ref': 'A', 'alt': 'T', 'cosmic_id': 'COSV101077077', 'mutation_desc': 'missense_variant'}\n",
      "{'gene_sym': 'PTPRD', 'chrom': '9', 'pos': 8460570, 'ref': 'T', 'alt': 'C', 'cosmic_id': 'COSV100645902', 'mutation_desc': 'missense_variant,splice_region_variant'}\n"
     ]
    }
   ],
   "source": [
    "matched_variants = []\n",
    "\n",
    "for variant in variants:\n",
    "    chrom = variant[\"chrom\"]\n",
    "    pos = variant[\"pos\"]\n",
    "    ref = variant[\"ref\"]\n",
    "    alt = variant[\"alt\"]\n",
    "        \n",
    "    query = \"\"\"\n",
    "        SELECT GENOMIC_MUTATION_ID, MUTATION_DESCRIPTION, GENE_SYMBOL\n",
    "        FROM cosmic_mutations\n",
    "        WHERE CHROMOSOME = %s\n",
    "        AND GENOME_START = %s\n",
    "        AND GENOMIC_WT_ALLELE = %s\n",
    "        AND GENOMIC_MUT_ALLELE = %s\n",
    "    \"\"\"\n",
    "    cursor.execute(query, (chrom, pos, ref, alt))\n",
    "    matching_rows = cursor.fetchall()\n",
    "\n",
    "    for row in matching_rows:\n",
    "        matched_variants.append({\n",
    "            \"gene_sym\":row[2],\n",
    "            \"chrom\": chrom,\n",
    "            \"pos\": pos,\n",
    "            \"ref\": ref,\n",
    "            \"alt\": alt,\n",
    "            \"cosmic_id\": row[0],\n",
    "            \"mutation_desc\": row[1]\n",
    "        })\n",
    "\n",
    "\n",
    "for variant in matched_variants:\n",
    "    print(variant)\n",
    "\n",
    "cursor.close()\n",
    "db_connection.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "oncotrack",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
