{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "from config import Config\n",
    "from sqlalchemy.orm import sessionmaker\n",
    "from sqlalchemy import create_engine\n",
    "import mysql.connector\n",
    "import pysam\n",
    "import pandas as pd\n",
    "# from model import db, Patient, Variant, Gene, patient_has_variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_connection = mysql.connector.connect(\n",
    "    host=\"localhost\",       # Cambia si estás usando un servidor remoto\n",
    "    user=\"root\",            # Tu usuario de MySQL\n",
    "    password=\"1234\",  # Cambia a tu contraseña\n",
    "    database=\"variantes_prueba\"  # Nombre de la base de datos donde tienes las variantes\n",
    ")\n",
    "cursor = db_connection.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GBM and LUAD don't show any mutation matching. \n",
    "vcf = pysam.VariantFile(\"/home/marti/MBHS/DBW/OncoTrack/vcf_mutations/AML_sample_muts.hg38.vcf\")\n",
    "variants = []\n",
    "for record in vcf:\n",
    "    chrom = record.chrom \n",
    "    pos = record.pos  \n",
    "    ref = record.ref  \n",
    "    \n",
    "    for alt in record.alts:\n",
    "        if len(ref) > 1 or len(alt) > 1:\n",
    "            ref = ref[1:] if len(ref) > 1 else \"\"\n",
    "            alt = alt[1:] if len(alt) > 1 else \"\"\n",
    "        variants.append({\n",
    "            \"chrom\": chrom,\n",
    "            \"pos\": pos,\n",
    "            \"ref\": ref,\n",
    "            \"alt\": alt  \n",
    "        })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_8842/1191615146.py:7: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('/home/marti/MBHS/DBW/OncoTrack/cosmut_db.tsv', sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1156314"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sqlalchemy import create_engine\n",
    "import pandas as pd\n",
    "\n",
    "# Conexión con MySQL\n",
    "# engine = create_engine('mysql+mysqlconnector://root:1234@localhost/variantes_prueba')\n",
    "\n",
    "# df = pd.read_csv('/home/marti/MBHS/DBW/OncoTrack/cosmut_db.tsv', sep='\\t')\n",
    "\n",
    "# df.to_sql('cosmic_mutations', con=engine, if_exists='replace', index=False)\n",
    "\n",
    "# CREATE INDEX idx_chromosome ON cosmic_mutations(CHROMOSOME(100));\n",
    "# CREATE INDEX idx_genome_start ON cosmic_mutations(GENOME_START);\n",
    "# CREATE INDEX idx_genomic_wt_allele ON cosmic_mutations(GENOMIC_WT_ALLELE(100)); \n",
    "# CREATE INDEX idx_genomic_mut_allele ON cosmic_mutations(GENOMIC_MUT_ALLELE(100));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'gene_sym': 'NRAS', 'chrom': '1', 'pos': 114713907, 'ref': 'T', 'alt': 'A', 'cosmic_id': 'COSV54736991', 'mutation_desc': 'missense_variant'}\n",
      "{'gene_sym': 'TP53', 'chrom': '17', 'pos': 7675994, 'ref': 'C', 'alt': 'T', 'cosmic_id': 'COSV52718605', 'mutation_desc': 'splice_region_variant,synonymous_variant'}\n"
     ]
    }
   ],
   "source": [
    "matched_variants = []\n",
    "\n",
    "for variant in variants:\n",
    "    chrom = variant[\"chrom\"]\n",
    "    pos = variant[\"pos\"]\n",
    "    ref = variant[\"ref\"]\n",
    "    alt = variant[\"alt\"]\n",
    "        \n",
    "    query = \"\"\"\n",
    "        SELECT variant_id, variant_type, gene\n",
    "        FROM variant\n",
    "        WHERE chromosome = %s\n",
    "        AND position = %s\n",
    "        AND reference = %s\n",
    "        AND alternative = %s\n",
    "    \"\"\"\n",
    "    cursor.execute(query, (chrom, pos, ref, alt))\n",
    "    matching_rows = cursor.fetchall()\n",
    "\n",
    "    for row in matching_rows:\n",
    "        matched_variants.append({\n",
    "            \"gene_sym\":row[2],\n",
    "            \"chrom\": chrom,\n",
    "            \"pos\": pos,\n",
    "            \"ref\": ref,\n",
    "            \"alt\": alt,\n",
    "            \"cosmic_id\": row[0],\n",
    "            \"mutation_desc\": row[1]\n",
    "        })\n",
    "\n",
    "for variant in matched_variants:\n",
    "    print(variant)\n",
    "\n",
    "cursor.close()\n",
    "db_connection.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "oncotrack",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
